client/components/sticky-panel/index.jsx

```diff
@@ -16,31 +16,151 @@ import afterLayoutFlush from 'lib/after-layout-flush';
+			</div>
+			<div className="sticky-panel__spacer" style={ { height: state.spacerHeight } } />
+		</div>
+	);
+}
+
+function isWindowTooSmall( minLimit ) {
+	return ( minLimit !== false && minLimit >= window.innerWidth ) || isMobile();
+}
+
+class StickyPanelWithIntersectionObserver extends React.Component {
+	static displayName = 'StickyPanel';
 
-	static defaultProps = {
-		minLimit: false,
+	static propTypes = commonPropTypes;
+	static defaultProps = commonDefaultProps;
+
+	state = {
+		isSticky: false,
+		spacerHeight: 0,
+		blockWidth: 0,
 	};
 
+	onIntersection = afterLayoutFlush( entries => {
+		if ( ! entries || ! entries[ 0 ] ) {
+			return;
+		}
+		const { intersectionRatio, rootBounds, boundingClientRect } = entries[ 0 ];
+		const isSticky = intersectionRatio < 1 && rootBounds.bottom > boundingClientRect.bottom;
+		this.updateStickyState( isSticky );
+	} );
+
+	throttleOnResize = throttle(
+		() => this.setState( prevState => getDimensions( this, prevState.isSticky ) ),
+		RESIZE_RATE_IN_MS
+	);
+
+	componentDidMount() {
+		window.addEventListener( 'resize', this.throttleOnResize );
+
+		this.deferredMount = defer( () => {
+			this.observer = new IntersectionObserver( this.onIntersection, {
+				threshold: [ 0, 1 ],
+				rootMargin: `-${ calculateOffset() }px 0px 0px 0px`,
+			} );
+			this.observer.observe( ReactDom.findDOMNode( this ) );
+		} );
+	}
+
+	componentWillUnmount() {
+		if ( this.observer ) {
+			this.observer.disconnect();
+		}
+		this.onIntersection.cancel();
+		clearTimeout( this.deferredMount );
+		window.removeEventListener( 'resize', this.throttleOnResize );
+	}
+
+	updateStickyState( isSticky ) {
+		if ( isWindowTooSmall( this.props.minLimit ) ) {
+			return this.setState( { isSticky: false } );
+		}
+
+		if ( isSticky !== this.state.isSticky ) {
+			this.setState( {
+				isSticky,
+				...getDimensions( this, isSticky ),
+			} );
+		}
+	}
+
+	render() {
+		return renderStickyPanel( this.props, this.state );
+	}
+}
+
+class StickyPanelWithScrollEvent extends React.Component {
+	static displayName = 'StickyPanel';
+
+	static propTypes = commonPropTypes;
+	static defaultProps = commonDefaultProps;
+
 	state = {
 		isSticky: false,
 		spacerHeight: 0,
 		blockWidth: 0,
 	};
 
+	onWindowScroll = afterLayoutFlush( () => {
+		// Determine vertical threshold from rendered element's offset relative the document
+		const threshold = ReactDom.findDOMNode( this ).getBoundingClientRect().top;
+		const isSticky = threshold < calculateOffset();
+		this.updateStickyState( isSticky );
+	} );
+
+	throttleOnResize = throttle(
+		() => this.setState( prevState => getDimensions( this, prevState.isSticky ) ),
+		RESIZE_RATE_IN_MS
+	);
+
 	componentDidMount() {
 		this.deferredTimer = defer( () => {
-			// Determine and cache vertical threshold from rendered element's
-			// offset relative the document
-			this.threshold = ReactDom.findDOMNode( this ).offsetTop;
-			this.updateIsSticky();
+			this.onWindowScroll();
```

Comment:
Could you just call `this.onWindowScroll()` here and get rid of the `deferredTimer` altogether? The `onWindowScroll` already uses `afterLayoutFlush` and therefore doesn't execute immediately. No need for one more `defer`.

Reply:
Removed!